src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs

```diff
@@ -0,0 +1,200 @@
+                        TokenLogMessages.IDX10514,
+                        keysAttempted?.ToString(),
+                        samlToken.Assertion.Signature.KeyInfo,
+                        GetErrorStrings(error, errors),
+                        samlToken),
+                    ValidationFailureType.SignatureValidationFailed,
+                    typeof(SecurityTokenInvalidSignatureException),
+                    ValidationError.GetCurrentStackFrame());
+
+            string? keysAttemptedString = null;
+            if (resolvedKey is not null)
+                keysAttemptedString = resolvedKey.ToString();
+            else if ((keysAttempted?.Length ?? 0) > 0)
+                keysAttemptedString = keysAttempted!.ToString();
+
+            if (keysAttemptedString is not null)
+                return new XmlValidationError(
+                    new MessageDetail(
+                        TokenLogMessages.IDX10512,
+                        keysAttemptedString,
+                        GetErrorStrings(error, errors),
+                        samlToken),
+                    ValidationFailureType.SignatureValidationFailed,
+                    typeof(SecurityTokenSignatureKeyNotFoundException),
+                    ValidationError.GetCurrentStackFrame());
+
+            return new XmlValidationError(
+                new MessageDetail(TokenLogMessages.IDX10500),
+                ValidationFailureType.SignatureValidationFailed,
+                typeof(SecurityTokenSignatureKeyNotFoundException),
+                ValidationError.GetCurrentStackFrame());
+        }
+
+        private static ValidationResult<SecurityKey> ValidateSignatureUsingKey(SecurityKey key, Saml2SecurityToken samlToken, ValidationParameters validationParameters, CallContext callContext)
+        {
+            ValidationResult<string> algorithmValidationResult = validationParameters.AlgorithmValidator(
+                        samlToken.Assertion.Signature.SignedInfo.SignatureMethod,
+                        key,
+                        samlToken,
+                        validationParameters,
+                        callContext);
+
+            if (!algorithmValidationResult.IsValid)
+            {
+                return algorithmValidationResult.UnwrapError().AddCurrentStackFrame();
+            }
+            else
+            {
+                var validationError = samlToken.Assertion.Signature.Verify(
+                key,
+                validationParameters.CryptoProviderFactory ?? key.CryptoProviderFactory,
+                callContext);
+
+                if (validationError is null)
+                {
+                    samlToken.SigningKey = key;
+
+                    return key;
+                }
+                else
+                {
+                    return validationError.AddCurrentStackFrame();
+                }
+            }
+        }
+
+        private static string GetErrorStrings(ValidationError? error, List<ValidationError>? errors)
```

Comment:
This seems like a copy of method in the previous file. Is there any reason we are duplicating the code.

Reply:
Agreed this could be common, nothing token specific about it.

Reply:
Yes, because the SAML2 code is almost a copy of the SAML one, but we are not in a position to refactor them into anything different at this time.

Reply:
@brentschmaltz unless we do some refactor to share code between SAML and SAML2, I don't see the point in introducing code sharing for one helper method.